apis/v1alpha1/route_types.go

```diff
@@ -35,6 +35,23 @@ const (
 	GatewayAllowSameNamespace GatewayAllowType = "SameNamespace"
 )
 
+const (
+	// AnnotationAppProtocol defines the protocol a Gateway should use for
+	// communication with a Kubernetes Service. The annotation serves as a stop-gap
+	// solution for the Kubernetes versions < 1.18, where the `AppProtocol` field
+	// on the `Service` resource is not available. Using this annotation is highly
+	// discouraged when AppProtocol field is available. The AppProtocol field, when
+	// populated, takes precedence over this annotation.
+	//
+	// The annotation value must be a comma separated with each value containing
+	// a port number followed by a forward slash and the protocol.
+	// Examples:
+	//
+	// - `networking.x-k8s.io/app-protocol: 8000/http,8443/https`
+	// - `networking.x-k8s.io/app-protocol: 4000/tcp,5000/tls`
```

Comment:
I think my thoughts from the previous PR still apply here. I think this would be cleaner on BackendPolicy where the field can simply refer to a single protocol, not a port/protocol combination like we'd need here.

Reply:
Could you elaborate a little on how this field in BackendPolicy work in k8s > 1.18? Will that field continue to exist forever? If yes, will it take precedence over AppProtocol?

Reply:
To clarify, I'm still suggesting an annotation, just on BackendPolicy. Initially I think the guidance should be that AppProtocol defined on the Service should be given precedence over this annotation. Maybe in the future we'll see more use cases here and we can graduate the annotation to a field. Or maybe most of our users will already be on k8s 1.18 and we won't need this annotation for other resources. Starting on BackendPolicy just seems like it gives us the most flexibility + the simplest config.

Reply:
> To clarify, I'm still suggesting an annotation, just on BackendPolicy.

Ah, I was understanding this all wrong.
Having an annotation on BackendPolicy that works similar to a field on the Service resource seems odd. It sits somewhere in the middle of "having a property in the BackendPolicy resource" and "having an annotation on the backend resource (i.e. Service in this case).

I do see the flexibility that it would bring, especially for non-Service backends. But then we need to make a decision if we expect AppProtocol to be the place for protocol (and then by extension, other resources which can be backends should also have a similar field) or is the protocol something that stays in BackendPolicy and that applies for all resource types.

I think we need to agree on the problem and the scope here. Are we providing a stop-gap solution for AppProtocol or do we see this as a larger problem and want to fix it properly?

Reply:
> To clarify, I'm still suggesting an annotation, just on BackendPolicy.

Ah, I was understanding this all wrong.
Having an annotation on BackendPolicy that works similar to a field on the Service resource seems odd. It sits somewhere in the middle of "having a property in the BackendPolicy resource" and "having an annotation on the backend resource (i.e. Service in this case).

I do see the flexibility that it would bring, especially for non-Service backends. But then we need to make a decision if we expect AppProtocol to be the place for protocol (and then by extension, other resources which can be backends should also have a similar field) or is the protocol something that stays in BackendPolicy and that applies for all resource types.

I think we need to agree on the problem and the scope here. Are we providing a stop-gap solution for AppProtocol or do we see this as a larger problem and want to fix it properly?